Client/Frontend/Home/BookmarksEditPanel.swift

```diff
@@ -0,0 +1,335 @@
+                    cell.textField.delegate = self
+                }
+            case .Folder:
+                cell.label.text = NSLocalizedString("Folder", comment: "Folder name of a bookmark.")
+                cell.textField.text = bookmarkFolder?.title ?? NSLocalizedString("Bookmarks", comment: "Panel accessibility label")
+                cell.textField.enabled = false
+            default:
+                break
+            }
+            return cell
+            
+        case 1:
+            return deleteButtonCell
+            
+        default:
+            // This should never happen.
+            return super.tableView(tableView, cellForRowAtIndexPath: indexPath)
+        }
+    }
+    
+    func textFieldDidChange(textField: UITextField) {
+        doneButton?.enabled = true
+    }
+    
+    func textFieldShouldReturn(textField: UITextField) -> Bool {
+        textField.resignFirstResponder()
+        return true
+    }
+    
+    func doneAndSave(button: UIButton) {
+        guard let source = self.source,
+              let bookmark = self.bookmark else {
+            return
+        }
+        
+        guard let factory = source.modelFactory.value.successValue else {
+            log.error("Couldn't get model factory. This is unexpected.")
+            self.onModelFailure(DatabaseError(description: "Unable to get factory."))
+            return
+        }
+        
+        if let err = factory.updateByGUID(BookmarkItem(guid: bookmark.guid, title: titleField?.text ?? bookmark.title, url: urlField?.text ?? bookmark.url)).value.failureValue {
+            log.debug("Failed to remove \(bookmark.guid).")
+            self.onModelFailure(err)
+            return
+        }
+        
+        NSNotificationCenter.defaultCenter().postNotificationName(BookmarkStatusChangedNotification, object: bookmark, userInfo:["added": false])
+        self.navigationController?.popViewControllerAnimated(true)
+    }
+    
+    override func tableView(tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+        if section != 0 {
+            return nil
+        }
+        
+        // Don't show a header for the root
+        if source == nil || parentFolders.isEmpty {
+            return nil
+        }
+        guard let header = tableView.dequeueReusableHeaderFooterViewWithIdentifier(BookmarkFolderEditHeaderViewIdentifier) as? BookmarkFolderEditTableViewHeader else { return nil }
+        
+        doneButton = header.doneButton
+        
+        header.doneButton.removeTarget(self, action: #selector(BookmarksEditPanel.doneAndSave(_:)), forControlEvents: UIControlEvents.TouchUpInside)
+        header.doneButton.addTarget(self, action: #selector(BookmarksEditPanel.doneAndSave(_:)), forControlEvents: UIControlEvents.TouchUpInside)
+        
+        // register as delegate to ensure we get notified when the user interacts with this header
+        if header.delegate == nil {
```

Comment:
Don't need this nil check - you can just go header.delegate = self. No harm in overwriting.